### PR TITLE
fix #44735, raise exception on `Vararg`-arguments in `subtype` rather than abort

### LINKED ISSUE #44735: typeintersect Vararg aborts
this is not good:
```
julia> typeintersect(Vararg{Int}, Int)
julia: /data/vtjnash/julia/src/subtype.c:1261: subtype: Assertion `!jl_is_vararg(x) && !jl_is_vararg(y)' failed.

signal (6): Aborted
```
and other mild oddities (I think these are correct, since they can appear as type parameters, they should be valid as arguments also)
```
julia> typeintersect(1, 2)
Union{}

julia> typeintersect(1, 1)
1

julia> typeintersect(1, Any)
1

julia> typeintersect(Any, 1)
1
```

### PR DESCRIPTION
**Pull request by aviatesk**

### CODE REVIEW

**Comment by aviatesk** [comment]
@vtjnash does this seem sensible to you?

**Comment by vtjnash** [comment]
I am not precisely sure how Vararg works now, but seems probably safe. Seems like something that should be handled internally to type-intersection though, where other conditions are checked?

**Comment by aviatesk** [comment]
yeah, if `jl_type_intersection` does really only expect `Type`-objects, I think it might be better to change the signature to `typeintersect(@nospecialize(a::Type), @nospecialize(b::Type))`, and remove the vararg assertion from `jl_type_intersection`?

**Comment by aviatesk** [comment]
ok, now the exception would be raised within subtype.c.

**Comment by Keno** [comment]
IIRC @JeffBezanson felt pretty strongly that passing non-type object here is a bug and should assert, but maybe I'm misremembering. Regardless, we should check with him when he's back from vacation.

**Comment by JeffBezanson** [comment]
Right, I don't think subtype/intersection should need to throw any errors. Currently there is only one (`circular type parameter constraint`), and it's basically a misplaced type-definition-time check.

I think we wanted this assert after the `TypeofVararg` change to make sure we caught all the cases that treated it like a type. Now that it's been working for a while, I think we could just remove the assert and let it fall through to `egal` like all other random values. If we later want to add checks that the arguments to `typeintersect` at the julia level are `Type`s, that might be ok too.

**Comment by aviatesk** [comment]
Thanks @JeffBezanson for your comment. I removed the `jl_is_vararg` assertions from subtype.c, and also added Julia level assertions that checks if input arguments are `Type`s.
I will try the `inference` benchmark: I will remove the Julia level error checks if benchmark result shows it can lead to some performance problem.

@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1089821608) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d39f6b_vs_2c8c0a4/report.md).

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1089838979) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4d39f6b_vs_2c8c0a4/report.md).

File: base/reflection.jl [hunk 1]
```diff
@@ -675,12 +675,17 @@ struct type with no fields.
...
+    typeintersect(T::Type, S::Type)
 
 Compute a type that contains the intersection of `T` and `S`. Usually this will be the
 smallest such type or one close to it.
 """
-typeintersect(@nospecialize(a), @nospecialize(b)) = (@_pure_meta; ccall(:jl_type_intersection, Any, (Any, Any), a, b))
+function typeintersect(@nospecialize(a), @nospecialize(b))
+    @_pure_meta
+    isa(a, Type) || throw(TypeError(:typeintersect, Type, typeof(a)))
+    isa(b, Type) || throw(TypeError(:typeintersect, Type, typeof(b)))
```
Comment (JeffBezanson) [inline base/reflection.jl:686]:
```suggestion
    isa(b, Type) || throw(TypeError(:typeintersect, Type, b))
```

File: base/reflection.jl [hunk 2]
```diff
@@ -675,12 +675,17 @@ struct type with no fields.
...
-    typeintersect(T, S)
+    typeintersect(T::Type, S::Type)
 
 Compute a type that contains the intersection of `T` and `S`. Usually this will be the
 smallest such type or one close to it.
 """
-typeintersect(@nospecialize(a), @nospecialize(b)) = (@_pure_meta; ccall(:jl_type_intersection, Any, (Any, Any), a, b))
+function typeintersect(@nospecialize(a), @nospecialize(b))
+    @_pure_meta
+    isa(a, Type) || throw(TypeError(:typeintersect, Type, typeof(a)))
```
[the remark this answers is not in the document]
Comment (vtjnash) [reply base/reflection.jl:685]: Why isn't this written `a::Type`?

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/44761#issuecomment-1091593989) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/4ed1360_vs_badad9d/report.md).

**Comment by aviatesk** [comment]
Seems good to go?

**Pull request merged**